File: prog-intro/HW 8 (PCMS)/J.java
package PCMS;

import java.util.*;

public class J {
    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt();
        int[][] array = new int[n][n];
        for (int i = 0; i < n; i++) {
            String str = in.next();
            for (int j = 0; j < n; j++) {
                array[i][j] = (Character.getNumericValue(str.charAt(j)));
            }
        }

        for (int i = 0; i < n - 2; i ++) {
            for (int j = i + 1; j < n - 1; j++) {
                if (array[i][j] > 0) {
                    for (int k = j + 1; k < n; k++) {
                        array[i][k] = (array[i][k] + 10 - array[j][k]) % 10;
                    }
                }
            }
        }

        for (int i = 0; i < n; i ++) {
            for (int j = 0; j < n; j++) {
                System.out.print(array[i][j]);
            }
            System.out.println();
        }

    }
}

File: algorithms-and-data-structures/07. DFS, MST/J.kt
import java.util.*
import kotlin.math.max
import kotlin.math.min


private lateinit var head: Array<Int>
private lateinit var dsu: Array<MutableList<Int>>
class J {
    fun main() {
        val scanner = Scanner(System.`in`)
        val n = scanner.nextInt()
        val m = scanner.nextInt()
        val Q = TreeSet(
            compareBy<Pair<Int, Pair<Int, Int>>> { it.first }.thenBy { it.second.first }.thenBy { it.second.second })

        head = Array(n) { i -> i }
        dsu = Array(n) { i -> listOf(i).toMutableList() }

        for (i in 0 until m) {
            val first = scanner.nextInt() - 1
            val second = scanner.nextInt() - 1
            val weight = scanner.nextInt()

            Q.add(Pair(weight, Pair(min(first, second), max(first, second))))
        }

        var summaryWeight = 0L

        for (x in Q) {
            val uv = x.second

            if (head[uv.first] != head[uv.second]) {
                union(uv.first, uv.second)
                summaryWeight += x.first
            }
        }

        println(summaryWeight)
    }

    fun union(first: Int, second: Int) {
        var x = head[first]
        var y = head[second]

        if (dsu[x].size > dsu[y].size) {
            val t = x
            x = y
            y = t
        }

        for (element in dsu[x]) {
            head[element] = y
            dsu[y].add(element)
        }

        dsu[x].clear()
    }
}